### ISSUE TITLE
In null safety mode, an error is thrown when the value of the map<string,string> is string empty

### ISSUE THREAD

**Issue by innwin** [google/protobuf.dart#508]
In null safety mode, an error is thrown when the value of the map<string,string> is string empty

![1621851174799](https://user-images.githubusercontent.com/6335833/119332871-c3834b80-bcbb-11eb-910c-654ad65a1a47.jpg)

![87511BED397FE585D3331D3587737962](https://user-images.githubusercontent.com/6335833/119332930-d7c74880-bcbb-11eb-929a-fe4c9c7835ec.jpg)

test proto: 

```proto
syntax = "proto3";

message MapTest {
  map<string,string> area = 5;
}
```

```dart
import 'dart:core' as $core;

import 'package:protobuf/protobuf.dart' as $pb;

class MapTest extends $pb.GeneratedMessage {
  static final $pb.BuilderInfo _i = $pb.BuilderInfo(const $core.bool.fromEnvironment('protobuf.omit_message_names') ? '' : 'MapTest', createEmptyInstance: create)
    ..m<$core.String, $core.String>(5, const $core.bool.fromEnvironment('protobuf.omit_field_names') ? '' : 'area', entryClassName: 'MapTest.AreaEntry', keyFieldType: $pb.PbFieldType.OS, valueFieldType: $pb.PbFieldType.OS)
    ..hasRequiredFields = false
  ;

  MapTest._() : super();
  factory MapTest({
    $core.Map<$core.String, $core.String>? area,
  }) {
    final _result = create();
    if (area != null) {
      _result.area.addAll(area);
    }
    return _result;
  }
  factory MapTest.fromBuffer($core.List<$core.int> i, [$pb.ExtensionRegistry r = $pb.ExtensionRegistry.EMPTY]) => create()..mergeFromBuffer(i, r);
  factory MapTest.fromJson($core.String i, [$pb.ExtensionRegistry r = $pb.ExtensionRegistry.EMPTY]) => create()..mergeFromJson(i, r);
  @$core.Deprecated(
  'Using this can add significant overhead to your binary. '
  'Use [GeneratedMessageGenericExtensions.deepCopy] instead. '
  'Will be removed in next major version')
  MapTest clone() => MapTest()..mergeFromMessage(this);
  @$core.Deprecated(
  'Using this can add significant overhead to your binary. '
  'Use [GeneratedMessageGenericExtensions.rebuild] instead. '
  'Will be removed in next major version')
  MapTest copyWith(void Function(MapTest) updates) => super.copyWith((message) => updates(message as MapTest)) as MapTest; // ignore: deprecated_member_use
  $pb.BuilderInfo get info_ => _i;
  @$core.pragma('dart2js:noInline')
  static MapTest create() => MapTest._();
  MapTest createEmptyInstance() => create();
  static $pb.PbList<MapTest> createRepeated() => $pb.PbList<MapTest>();
  @$core.pragma('dart2js:noInline')
  static MapTest getDefault() => _defaultInstance ??= $pb.GeneratedMessage.$_defaultFor<MapTest>(create);
  static MapTest? _defaultInstance;

  @$pb.TagNumber(5)
  $core.Map<$core.String, $core.String> get area => $_getMap(0);
}
```

test code: 
```dart
import 'package:test/test.dart';

import 'maptest.pb.dart';

void main() {
  test('map test', () {
    final b = [42, 4, 10, 2, 99, 110, 42, 4, 10, 2, 104, 107, 42, 8, 10, 2, 101, 110, 18, 2, 78, 79];
    final mapTest = MapTest.fromBuffer(b);
    print(mapTest);
  });
}
```

**Comment by daanporon** [comment]
In our codebase we are experiencing the same issue. We also have a map<string, string> and from the moment the value is an empty string we receive a DATA_LOSS exception. Currently we are still not using null-safety, so i don't think it's related to that.

**Comment by LeFrosch** [comment]
Our application uses a C# backend and we use a `map<uint32, string>`. Everything works as expected; only `0` can not be used as a key. When the dart client cannot decode such a map and throws a `DATA_LOSS` exception.

**Comment by JimHume** [comment]
The code from the PR has been tested locally and has proven to resolve the DATA_LOSS issue with header-loading, when the value of a header is an empty string.

The C# server sets a handful of properties on the header as part of initiating a stream, and when said stream initiation hits the dart side it is crashing with "DATA_LOSS" if one of said header properties was set to `string.Empty` on the C# server. This somehow gets interpreted as NULL and throws the "DATA_LOSS" exception.

Is this only an issue with headers? Will setting regular message properties to `string.Empty` in C# result in similar exceptions being thrown in dart when the message is received by a dart client?

Or is this only an issue with map fields, as indicated by most responders to this issue?

TL;DR: the code from the PR that @sigurdm [provided](https://github.com/google/protobuf.dart/pull/536) was tested locally and does indeed solve this issue--can this please be approved?
